src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,28 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (_, &ty::TyRef(_, checked)) => {
+                if self.infcx.can_sub(self.param_env, checked.ty, &expected).is_ok() {
+                    match expr.node {
+                        hir::ExprAddrOf(_, ref expr) => {
+                            if let Ok(code) = self.tcx.sess.codemap().span_to_snippet(expr.span) {
+                                return Some(format!("try with `{}`", code));
+                            }
+                        }
+                        _ => {
+                            if self.infcx.type_moves_by_default(self.param_env,
```

Comment:
@nikomatsakis: This call is never returning true so I suppose I'm misusing it. At first, I checked on `checked_ty` directly but then I thought that the fact it was a reference might make it not work. Then I tried on the lower level (which is the current code) but still can't make it work. Any idea what I'm doing wrong?

Reply:
I think you want `!self.infcx.type_moves_by_default` -- that is, this function returns `false` is the value is `Copy` (and hence it does *not* move).